sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2AlterTableCommands.scala

```diff
@@ -295,7 +295,16 @@ case class AlterTableCollation(
 case class AddConstraint(
     table: LogicalPlan,
     tableConstraint: TableConstraint) extends AlterTableCommand {
-  override def changes: Seq[TableChange] = Seq.empty
+  override def changes: Seq[TableChange] = {
+    val constraint = tableConstraint.toV2Constraint(isCreateTable = false)
+    val validatedTableVersion = table match {
+      case t: ResolvedTable if constraint.enforced() =>
+        t.table.currentVersion()
+      case _ =>
+        null
+    }
+    Seq(TableChange.addConstraint(constraint, validatedTableVersion))
```

Comment:
CHECK constraints must optionally validate existing data. Am I right this PR doesn't have this? What would be our plan?

Reply:
> must optionally validate

Make sense. Do you mean `CHECK ... NOT ENFOCED`?

Reply:
ENFORCED/NOT ENFORCED impacts subsequent writes. I was referring to ALTER TABLE ... ADD CONSTRAINT that must scan the existing data.

Reply:
Just for my understanding. Anton's comment was about how to validate the existing data in `ALTER TABLE ... ADD CONSTRAINT`. Is it addressed in this PR, @gengliangwang ?

The above follow-up JIRA (SPARK-51905) is not about that, isn't it?

> [SPARK-51905](https://issues.apache.org/jira/browse/SPARK-51905) Disallow NOT ENFORCED CHECK constraint

Reply:
Actually, @gengliangwang already created it: https://issues.apache.org/jira/browse/SPARK-51903.

Reply:
Got it. Thank you, @aokolnychyi . Ya, SPARK-51903 is what I expected.